pkg/reconciler/managed/reconciler.go

```diff
@@ -316,12 +326,19 @@ type ExternalObservation struct {
 
 // An ExternalCreation is the result of the creation of an external resource.
 type ExternalCreation struct {
-	// ExternalNameAssigned is true if the Create operation resulted in a change
-	// in the external name annotation. If that's the case, we need to issue a
-	// spec update and make sure it goes through so that we don't lose the identifier
-	// of the resource we just created.
+	// ExternalNameAssigned should be true if the Create operation resulted
+	// in a change in the resource's external name. This is typically only
+	// needed for external resource's with unpredictable external names that
+	// are returned from the API at create time.
 	ExternalNameAssigned bool
 
+	// ExternalCreateTimeSet should be true if the Create operation resulted
+	// in the resource's external create time being set. This is typically
+	// only needed to accommodate APIs that have some delay between when an
+	// external resource is created and when the external resource is
+	// reported to exist.
+	ExternalCreateTimeSet bool
```

Comment:
In retrospect it might have been nicer for this API to return these values rather than signal that they'd been set, but since we've established a pattern with `ExternalNameAssigned` I figure we may as well stick with it.

Reply:
Hrm - another approach here would be to just always set the `crossplane.io/external-create-time` annotation in the `managed.Reconciler` rather than relying on `Create` implementations to optionally set it. I could imagine the time at which the external resource was created being of some use to folks even in a purely informational sense for debugging etc.

I think the only downside here would be that we'd always `Update` the managed resource immediately after calling `Create`, which resets any pending managed resource status updates. Today the `Update` is effectively opt-in, though to be fair these implications aren't really obvious to `ExternalClient` authors as is.

Reply:
I would agree with this approach.

Regarding the downside, couldn't we fix that by storing status to a temp var before running `Update()` and then run a `Status.Update()` using the stored status?

Reply:
> Regarding the downside, I believe we could fix that by storing status to a temp var before running Update() and then run a Status.Update() using the stored status?

Proooobably - the challenge would be extracting it. The `managed.Reconciler` deals in the `resource.Managed` interface, which has no awareness of the `Spec` and `Status` of the type it contains. Off the top of my head the only way to do this would be to convert to an `*Unstructured`, which I'm a little hesitant to do for such a core library.

I'd perhaps lean toward just advising folks that they should not update the `Status` during the `Create` call. I believe the existing convention is to modify as little of the managed resource as possible during `Create`, then use `Observe` to set `Status` and late-init `Spec`.

Reply:
I remember that I've dealt with this situation by:
1. Taking a deepcopy of the object with the desired status, 
2. Do the `Update`
3. Set resource version returned by update call
4. and do a status update. 

Something like:

```
++ mgWithDesiredStatus := managed.DeepCopyObject()

err := retry.OnError(retry.DefaultRetry, resource.IsAPIError, func() error {
	nn := types.NamespacedName{Name: managed.GetName()}
	if err := r.client.Get(ctx, nn, managed); err != nil {
		return err
	}
	meta.SetExternalName(managed, en)
	return r.client.Update(ctx, managed)
})
if err != nil {
	log.Debug("Cannot update managed resource", "error", err)
	record.Event(managed, event.Warning(reasonCannotUpdateManaged, errors.Wrap(err, errUpdateManagedAfterCreate)))
	managed.SetConditions(xpv1.ReconcileError(errors.Wrap(err, errUpdateManagedAfterCreate)))
	return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
}

++ mgWithDesiredStatus.(metav1.Object).SetResourceVersion(managed.GetResourceVersion())
++ r.client.Status().Update(ctx, mgWithDesiredStatus)
```

Feels a bit hacky, but worked well so far :)

Reply:
This sounds a little like what our "updating" applicator does. I think the only risk there is that copying the resource version defeats the API server's concurrency control (i.e. we'd blindly overwrite anything that happened between our deepcopy and our status update). I suppose that's not a _huge_ issue when it comes to managed resource status, given we don't really expect write contention with other controllers on the status field, and we also don't rely on any state being persisted to that object.

Reply:
I've pushed a commit that moves responsibility for setting the `crossplane.io/external-create-time` annotation into the `managed.Reconciler`, and thus updates annotations after every successful create. For the time being I've opted not to worry about persisting status. It occurred to me that the current implementation actually discards _all_ updates made during `Create` with the exception of annotation changes.

To my knowledge we don't have any `Create` implementations that expected spec or status updates to be persisted automatically, so I've documented this expectation.